ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs

```diff
@@ -216,6 +232,53 @@ data Mempool m blk idx = Mempool {
     , zeroIdx        :: idx
     }
 
+-- | The result of attempting to add a transaction to the mempool.
+data MempoolAddTxResult blk
+  = TxAdded
+    -- ^ The transaction was added to the mempool.
+  | TxRejected !(MempoolAddTxError blk)
+    -- ^ The transaction was rejected and could not be added to the mempool
+    -- for the specified reason.
+
+isTxAdded :: MempoolAddTxResult blk -> Bool
+isTxAdded TxAdded = True
+isTxAdded _       = False
+
+isTxRejected :: MempoolAddTxResult blk -> Bool
+isTxRejected (TxRejected _) = True
+isTxRejected _              = False
+
+-- | An error encountered when attempting to add a transaction to the mempool.
+data MempoolAddTxError blk
```

Comment:
The type Edsko and I suggested in https://github.com/input-output-hk/ouroboros-network/issues/1161#issuecomment-594423279 intentionally did not classify "already in Mempool" as an error, as it's not really an error. For [example](https://github.com/input-output-hk/ouroboros-network/issues/1161#issuecomment-595093203), the wallet will resubmit transactions that didn't make it into the last block via the `LocalTxSubmissionServer`. The wallet doesn't know that the transaction is still in the Mempool, so we shouldn't treat that case as an error (think of it as a trace message severity). So go for something like:

```haskell
data MempoolAddTxResult blk
  = MempoolTxAdded
  | MempoolTxRejected (ApplyTxErr blk)
  | MempoolTxAlreadyInMempool
```

Also important: this new `MempoolAddTxError` type replaces the existing `ApplyTxError` type in the `LocalTxSubmission` protocol codec, so this breaks network compatibility, which we have to plan carefully. The 3-constructor approach doesn't break network compatibility.